### ISSUE TITLE
Android Cordova export broken

### ISSUE THREAD

**Issue by Ldwebster111** [Scirra/Construct-3-bugs#3094]
## Problem description

Exports using Android Cordova APK is no longer working. Noticed this last night when I tried to roll out a new version of my project to the Play Store. When launching through Play Store, game installs but instantly crashes. Trying to install a standalone unsigned APK using package installer means that the install won't even complete, stating that the file is corrupt.

At first I thought this might be related to the beta channel, so I opened an older version of my project that had successfully exported previously, using the stable channel, and the same problem persists.

To double check that it wasn't something in my project, I created a new, blank project using the stable release, added a single sprite and the bare minimum project description to get the Cordova export to work, and tested that as an unsigned APK. As before, the installer claims that the file is corrupted. Attached .c3p for testing purposes, though it's basically an empty project.

Also tested using C2 runtime and still getting same issue.

edit - just to be thorough, tried this on a separate C3 account and over a different internet network. Again, same result.

## Attach a .c3p

[Android_export_test.zip](https://github.com/Scirra/Construct-3-bugs/files/3342447/Android_export_test.zip)

## Steps to reproduce

1. export project using Android Cordova unsigned APK
2. install to Android phone using package installer

## Observed result

Corrupted file warning for unsigned APK. Failure to launch if signed and uploaded to Play Store.

## Expected result

To install and launch

## More details

Tested on two separate PCs, in both stable and beta channels. 

**Affected browsers/platforms:** Chrome/Android

first noticed in r156.2 though as mentioned, affects earlier builds.

**Comment by shortercode** [comment]
Mmm that's odd... I don't think I've changed anything which could cause
this. But the package installer is quite brittle. The corruption message
covers any number of failures and makes it quite awkward to find out the
real reason. If you attach logcat to the device during the install and look
at it's unfiltered logs then there's normally a reason printed, but it can
be quite hard to see around all the other messages.

I tried a quick debug build of space blaster on my phone and didn't see any
issues with installation. So this might not be affecting everyone. I'll do
some more testing when I get into the office on Monday, but any further
investigation you can do would be appreciated.

**Comment by Ldwebster111** [comment]
okay, looks like I'm an idiot and forgot to test on a second phone. Grabbed my wife's phone (an S7 like mine) and tested it on that and it's fine, so it's specific to my phone. I have no idea what's causing it though, everything else runs fine and I haven't installed anything on it recently.

I might try a software update and see if that does anything. Will also test on a few workmate's phones tomorrow to confirm it's just mine.

Sorry for wasting your time.

**Comment by shortercode** [comment]
I have come across issues like this which are specific to certain devices.
For example: some devices won't install apps with package IDs that include
uppercase characters.

**Comment by Ldwebster111** [comment]
hmmm, I just tested on my son's S4 and it won't launch on his either. I double-checked my wife's and it's the same version. Not sure what's going on here.

**Comment by Ldwebster111** [comment]
I need to go to bed as I have work in the morning. I keep thinking about what I could have done to my project to break it and can't think of anything, but then I remember that I can't get a blank APK to install either, so I can't see it being related to this specific project. I'll keep testing things tomorrow, but I'm starting to get worried as I'm being assessed for a government grant this week and the last thing I need is for them to test the game and find it won't launch.

**Comment by Ldwebster111** [comment]
Okay, I'm totally going to bed now. But I checked the stack trace from the error log and it says I have an invalid application ID. Is this related to adMob?

**Comment by shortercode** [comment]
Yes there's been some changes to the Admob SDK, it's now required to
specify your app ID at compile time. If the ID is invalid then it will
crash the app when you try to start it.

The value is tied to the Android app ID property of the plugin in the
current beta releases.

I wouldn't expect this to cause installation issues though. Take another
look at the log in the morning, dealing with software problems when your
tired is a recipe for stress and disaster.

**Comment by Ldwebster111** [comment]
I'm at work now and just did a test. Grabbed the Space Blaster template and exported it as an unsigned APK file. Sent the file to a student's phone (Google Pixel 2) and got them to install it. Again, it wouldn't install.

Here's the link if you want to try it. I haven't added a single thing to this project, just opened it and exported as unsigned.

https://drive.google.com/open?id=1WAGRFVq6534tBANV_6RqzTmY1Q4WqcxC

**Comment by Ldwebster111** [comment]
Uploaded a version of my game to Google Play with all Admob and MobileIAP events removed and it worked. Guessing I have something configured incorrectly. At least I've narrowed down where to look.

I wonder if the testing of unsigned APKs isn't a different matter entirely.

**Comment by Ldwebster111** [comment]
> Yes there's been some changes to the Admob SDK, it's now required to
specify your app ID at compile time. If the ID is invalid then it will
crash the app when you try to start it.

>The value is tied to the Android app ID property of the plugin in the
current beta releases.

So this was the problem. I had missed the admob properties app ID before and, tbh, had to read what you were saying a couple of times before it clicked what I was looking for. Seems to be fixed now in that my game launches through Google Play again (phew). Still not sure why the unsigned APKs wouldn't launch, but I'm guessing that's a separate issue.

Cheers

**Comment by shortercode** [comment]
Yeah I think I need to do some sort of PSA for the Admob plugin when the stable goes out. I have an idea on how to improve the export process to make things easier for users. But the changes I need to make are larger than I would want to put in at this point in the beta cycle. So they will have to wait we start the next unfortunately.

I think you had 2 issues here, as you've found. We can consider the Admob stuff closed if you've got it working. 

As for the APK stuff that's a little murkier... You say the APK you installed on the students phone was an unsigned release, I wouldn't expect that to install on a phone anyway. I believe debug APKs are signed with a special development certificate, which devices can recognise.

**Comment by Ldwebster111** [comment]
Aw man, I think you're right. It's been ages since I exported using anything other than a signed APK, so for some reason I was using unsigned instead of debug. As soon as a read that I knew I'd screwed up.

Definitely think a PSA would be a good idea. I spent so long on this when in reality, it was a 2 minute fix.

I think this can be closed. It was a combination of me missing some basic information and also being a bit of a dope.

Thanks
